doc/neps/nep-0037-array-module.rst

```diff
@@ -0,0 +1,332 @@
+
+The actual implementation of `get_array_module` will be in C, but should be equivalent to this Python code:
+
+.. code::python
+
+    def get_array_module(*arrays, default=numpy):
+        arrays, types = _implementing_arrays_and_types(arrays)
+        if not implementing_arrays and default is not None:
+            return default
+        for array in implementing_arrays:
+            module = array.__array_module__(types)
+            if module is not NotImplemented:
+                return module
+        raise TypeError("no common array module found")
+
+    def _implementing_arrays_and_types(relevant_arrays):
+        types = []
+        implementing_arrays = []
+        for array in arrays:
+            t = type(array)
+            if t not in types and hasattr(t, '__array_module__'):
+                types.append(t)
+                implementing_arrays.append(array)
+        return implementing_arrays, types
+
+Relationship with ``__array_ufunc__`` and ``__array_function__``
+---------------------------------------------------------------
+
+These older protocols have distinct use-cases and should remain
+===============================================================
+
+``__array_module__`` is intended to resolve limitations of ``__array_function__``, so it is natural to consider whether it could entirely replace ``__array_function__``. This would offer dual benefits: (1) simplifying the user-story about how to override NumPy and (2) removing the slowdown associated with checking for dispatch when calling every NumPy function.
+
+However, ``__array_module__`` and ``__array_function__`` are pretty different from a user perspective: it requires explicit calls to ``get_array_function``, rather than simply reusing original ``numpy`` functions. This is probably fine for *libraries* that rely on duck-arrays, but may be frustratingly verbose for interactive use.
+
+Some of the dispatching use-cases for ``__array_ufunc__`` are also solved by ``__array_module__``, but not all of them. For example, it is still useful to be able to define non-NumPy ufuncs (e.g., from Numba or SciPy) in a generic way on non-NumPy arrays (e.g., with dask.array).
+
+Given their existing adoption and distinct use cases, we don't think it makes sense to remove or deprecate ``__array_function__`` and ``__array_ufunc__`` at this time.
+
+Mixin classes to implement ``__array_function__`` and ``__array_ufunc__``
+=========================================================================
+
+Despite the user-facing differences, ``__array_module__`` and a module implementing NumPy's API still contain sufficient functionality needed to implement dispatching with the existing duck array protocols.
+
+To make it easier to write duck arrays, we plan to add mixin classes (into ``numpy.lib.mixins``) that provide sensible defaults for these special methods in terms of ``get_array_module`` and ``__array_module__``, e.g.,
```

Comment:
Seems OK to me, but if anyone is unsure about it, it would probably OK to just have the code sample.